<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# Exploring and pre-processing a dataset using Pandas 


Estimated time needed: **30** minutes
    

## Objectives

After completing this lab you will be able to:

* Explore the dataset
* Pre-process dataset as required (may be for visualization)


## Introduction

The aim of this lab is to provide you a refresher on the **Pandas** library, so that you can pre-process and anlyse the datasets before applying data visualization techniques on it. This lab will work as acrash course on *pandas*. if you are interested in learning more about the *pandas* library, detailed description and explanation of how to use it and how to clean, munge, and process data stored in a *pandas* dataframe are provided in our course [**Data Analysis with Python**](https://www.coursera.org/learn/data-analysis-with-python?specialization=ibm-data-analyst) and [**Python for Applied Data Science**](https://www.coursera.org/learn/python-for-applied-data-science-ai?specialization=ibm-data-analyst)

------------


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Exploring Datasets with *pandas*](#0)<br>
2. [The Dataset: Immigration to Canada from 1980 to 2013](#1)<br>
3. [*pandas* Basics](#2) <br>
4. [*pandas* Intermediate: Indexing and Selection](#3) <br>
5. [*pandas* Filtering based on a criteria](#4)<br>
6. [*pandas* Sorting Values](#5)

</div>


# Exploring Datasets with *pandas* <a id="0"></a>

*pandas* is an essential data analysis toolkit for Python. From their [website](http://pandas.pydata.org/):
>*pandas* is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, **real world** data analysis in Python.

The course heavily relies on *pandas* for data wrangling, analysis, and visualization. We encourage you to spend some time and familiarize yourself with the *pandas* API Reference: http://pandas.pydata.org/pandas-docs/stable/api.html.


## The Dataset: Immigration to Canada from 1980 to 2013 <a id="1"></a>


Dataset Source: [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/development/desa/pd/data/international-migration-flows).

The dataset contains annual data on the flows of international immigrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. The current version presents data pertaining to 45 countries.

In this lab, we will focus on the Canadian immigration data.

![Data Preview](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png)

 The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">here</a>.

---


## *pandas* Basics<a id="2"></a>


The first thing we'll do is install **openpyxl** (formerly **xlrd**), a module that *pandas* requires to read Excel files.


In [ ]:
!mamba install openpyxl==3.0.9 -y

Next, we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [2]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library
import requests

In [6]:
# URL of the Excel file
file_url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx'

# Send a GET request to the URL
response = requests.get(file_url)

# Check if the request was successful (status code 200)
if response.status_code == 200:
    # Open a local file in binary write mode
    with open('canada.xlsx', 'wb') as f:
        # Write the content of the response to the local file
        f.write(response.content)
    print("File downloaded successfully.")
else:
    print("Failed to download the file.")

File downloaded successfully.


Let's download and import our primary Canadian Immigration dataset using *pandas*'s `read_excel()` method.


In [71]:
df_can = pd.read_excel(
    "canada.xlsx",
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2)

print('Data read into a pandas dataframe!')

Data read into a pandas dataframe!


Let's view the top 5 rows of the dataset using the `head()` function.


In [51]:
df_can.head()
# tip: You can specify the number of rows you'd like to see as follows: df_can.head(10) 

Type    Coverage          OdName  AREA AreaName   REG  \
0  Immigrants  Foreigners     Afghanistan   935     Asia  5501   
1  Immigrants  Foreigners         Albania   908   Europe   925   
2  Immigrants  Foreigners         Algeria   903   Africa   912   
3  Immigrants  Foreigners  American Samoa   909  Oceania   957   
4  Immigrants  Foreigners         Andorra   908   Europe   925   

           RegName  DEV             DevName  1980  ...  2004  2005  2006  \
0    Southern Asia  902  Developing regions    16  ...  2978  3436  3009   
1  Southern Europe  901   Developed regions     1  ...  1450  1223   856   
2  Northern Africa  902  Developing regions    80  ...  3616  3626  4807   
3        Polynesia  902  Developing regions     0  ...     0     0     1   
4  Southern Europe  901   Developed regions     0  ...     0     0     1   

   2007  2008  2009  2010  2011  2012  2013  
0  2652  2111  1746  1758  2203  2635  2004  
1   702   560   716   561   539   620   603  
2  3623  4005  5393  4752  4325  3774  4331  
3     0     0     0     0     0     0     0  
4     1     0     0     0     0     1     1  

[5 rows x 43 columns]

We can also view the bottom 5 rows of the dataset using the `tail()` function.


In [11]:
df_can.tail()

Type    Coverage          OdName  AREA AreaName  REG  \
190  Immigrants  Foreigners        Viet Nam   935     Asia  920   
191  Immigrants  Foreigners  Western Sahara   903   Africa  912   
192  Immigrants  Foreigners           Yemen   935     Asia  922   
193  Immigrants  Foreigners          Zambia   903   Africa  910   
194  Immigrants  Foreigners        Zimbabwe   903   Africa  910   

                RegName  DEV             DevName  1980  ...  2004  2005  2006  \
190  South-Eastern Asia  902  Developing regions  1191  ...  1816  1852  3153   
191     Northern Africa  902  Developing regions     0  ...     0     0     1   
192        Western Asia  902  Developing regions     1  ...   124   161   140   
193      Eastern Africa  902  Developing regions    11  ...    56    91    77   
194      Eastern Africa  902  Developing regions    72  ...  1450   615   454   

     2007  2008  2009  2010  2011  2012  2013  
190  2574  1784  2171  1942  1723  1731  2112  
191     0     0     0     0     0     0     0  
192   122   133   128   211   160   174   217  
193    71    64    60   102    69    46    59  
194   663   611   508   494   434   437   407  

[5 rows x 43 columns]

When analyzing a dataset, it's always a good idea to start by getting basic information about your dataframe. We can do this by using the `info()` method.

This method can be used to get a short summary of the dataframe.


In [14]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195 entries, 0 to 194
Columns: 43 entries, Type to 2013
dtypes: int64(37), object(6)
memory usage: 65.6+ KB


To get the list of column headers we can call upon the data frame's `columns` instance variable.


In [15]:
df_can.columns

Index([    'Type', 'Coverage',   'OdName',     'AREA', 'AreaName',      'REG',
        'RegName',      'DEV',  'DevName',       1980,       1981,       1982,
             1983,       1984,       1985,       1986,       1987,       1988,
             1989,       1990,       1991,       1992,       1993,       1994,
             1995,       1996,       1997,       1998,       1999,       2000,
             2001,       2002,       2003,       2004,       2005,       2006,
             2007,       2008,       2009,       2010,       2011,       2012,
             2013],
      dtype='object')

Similarly, to get the list of indices we use the `.index` instance variables.


In [16]:
df_can.index

RangeIndex(start=0, stop=195, step=1)

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [17]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [18]:
df_can.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013]

In [19]:
df_can.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [20]:
print(type(df_can.columns.tolist()))
print(type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


In [21]:
# size of dataframe (rows, columns)
df_can.shape

(195, 43)

**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [72]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions    16    39   
1      Albania   Europe  Southern Europe   Developed regions     1     0   

   1982  1983  1984  1985  ...  2004  2005  2006  2007  2008  2009  2010  \
0    39    47    71   340  ...  2978  3436  3009  2652  2111  1746  1758   
1     0     0     0     0  ...  1450  1223   856   702   560   716   561   

   2011  2012  2013  
0  2203  2635  2004  
1   539   620   603  

[2 rows x 38 columns]

Let's rename the columns so that they make sense. We can use `rename()` method by passing in a dictionary of old and new names as follows:


In [73]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.dtypes

Country      object
Continent    object
Region       object
DevName      object
1980          int64
1981          int64
1982          int64
1983          int64
1984          int64
1985          int64
1986          int64
1987          int64
1988          int64
1989          int64
1990          int64
1991          int64
1992          int64
1993          int64
1994          int64
1995          int64
1996          int64
1997          int64
1998          int64
1999          int64
2000          int64
2001          int64
2002          int64
2003          int64
2004          int64
2005          int64
2006          int64
2007          int64
2008          int64
2009          int64
2010          int64
2011          int64
2012          int64
2013          int64
dtype: object

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [74]:
df_can['Total'] = df_can.iloc[:, 4:].sum(axis=1)
df_can["Total"]

0      58639
1      15699
2      69439
3          6
4         15
       ...  
190    97146
191        2
192     2985
193     1677
194     8598
Name: Total, Length: 195, dtype: int64

We can check to see how many null objects we have in the dataset as follows:


In [75]:
df_can.isnull().sum()

Country      0
Continent    0
Region       0
DevName      0
1980         0
1981         0
1982         0
1983         0
1984         0
1985         0
1986         0
1987         0
1988         0
1989         0
1990         0
1991         0
1992         0
1993         0
1994         0
1995         0
1996         0
1997         0
1998         0
1999         0
2000         0
2001         0
2002         0
2003         0
2004         0
2005         0
2006         0
2007         0
2008         0
2009         0
2010         0
2011         0
2012         0
2013         0
Total        0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


In [47]:
df_can.describe()

1980          1981          1982          1983          1984  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean     508.394872    566.989744    534.723077    387.435897    376.497436   
std     1949.588546   2152.643752   1866.997511   1204.333597   1198.246371   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%       13.000000     10.000000     11.000000     12.000000     13.000000   
75%      251.500000    295.500000    275.000000    173.000000    181.000000   
max    22045.000000  24796.000000  20620.000000  10015.000000  10170.000000   

              1985         1986          1987          1988          1989  \
count   195.000000   195.000000    195.000000    195.000000    195.000000   
mean    358.861538   441.271795    691.133333    714.389744    843.241026   
std    1079.309600  1225.576630   2109.205607   2443.606788   2555.048874   
min       0.000000     0.000000      0.000000      0.000000      0.000000   
25%       0.000000     0.500000      0.500000      1.000000      1.000000   
50%      17.000000    18.000000     26.000000     34.000000     44.000000   
75%     197.000000   254.000000    434.000000    409.000000    508.500000   
max    9564.000000  9470.000000  21337.000000  27359.000000  23795.000000   

       ...          2005          2006          2007          2008  \
count  ...    195.000000    195.000000    195.000000    195.000000   
mean   ...   1320.292308   1266.958974   1191.820513   1246.394872   
std    ...   4425.957828   3926.717747   3443.542409   3694.573544   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...     28.500000     25.000000     31.000000     31.000000   
50%    ...    210.000000    218.000000    198.000000    205.000000   
75%    ...    832.000000    842.000000    899.000000    934.500000   
max    ...  42584.000000  33848.000000  28742.000000  30037.000000   

               2009          2010          2011          2012          2013  \
count    195.000000    195.000000    195.000000    195.000000    195.000000   
mean    1275.733333   1420.287179   1262.533333   1313.958974   1320.702564   
std     3829.630424   4462.946328   4030.084313   4247.555161   4237.951988   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%       36.000000     40.500000     37.500000     42.500000     45.000000   
50%      214.000000    211.000000    179.000000    233.000000    213.000000   
75%      888.000000    932.000000    772.000000    783.000000    796.000000   
max    29622.000000  38617.000000  36765.000000  34315.000000  34129.000000   

              Total  
count  1.950000e+02  
mean   6.522651e+04  
std    1.820699e+05  
min    2.000000e+00  
25%    1.903000e+03  
50%    9.825000e+03  
75%    4.356900e+04  
max    1.374928e+06  

[8 rows x 35 columns]

---
## *pandas* Intermediate: Indexing and Selection (slicing)<a id="3"></a>


### Select Column
**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.
```python
    df.column_name               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python 
    df[['column 1', 'column 2']]  # returns dataframe
```
---


Example: Let's try filtering on the list of countries ('Country').


In [76]:
df_can.Country  # returns a series

0         Afghanistan
1             Albania
2             Algeria
3      American Samoa
4             Andorra
            ...      
190          Viet Nam
191    Western Sahara
192             Yemen
193            Zambia
194          Zimbabwe
Name: Country, Length: 195, dtype: object

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [77]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers. 
# for the sake of consistency, we will convert all column names to string later on.

Country  1980  1981  1982  1983  1984  1985
0       Afghanistan    16    39    39    47    71   340
1           Albania     1     0     0     0     0     0
2           Algeria    80    67    71    69    63    44
3    American Samoa     0     1     0     0     0     0
4           Andorra     0     0     0     0     0     0
..              ...   ...   ...   ...   ...   ...   ...
190        Viet Nam  1191  1829  2162  3404  7583  5907
191  Western Sahara     0     0     0     0     0     0
192           Yemen     1     2     1     6     0    18
193          Zambia    11    17    11     7    16     9
194        Zimbabwe    72   114   102    44    32    29

[195 rows x 7 columns]

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [33]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [78]:
df_can.head(3)

Country Continent           Region             DevName  1980  1981  \
0  Afghanistan      Asia    Southern Asia  Developing regions    16    39   
1      Albania    Europe  Southern Europe   Developed regions     1     0   
2      Algeria    Africa  Northern Africa  Developing regions    80    67   

   1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
0    39    47    71   340  ...  3436  3009  2652  2111  1746  1758  2203   
1     0     0     0     0  ...  1223   856   702   560   716   561   539   
2    71    69    63    44  ...  3626  4807  3623  4005  5393  4752  4325   

   2012  2013  Total  
0  2635  2004  58639  
1   620   603  15699  
2  3774  4331  69439  

[3 rows x 39 columns]

In [41]:
# optional: to remove the name of the index
df_can.index.name = None

df_can.reset_index()

index Continent              Region             DevName  1980  \
0       Afghanistan      Asia       Southern Asia  Developing regions    16   
1           Albania    Europe     Southern Europe   Developed regions     1   
2           Algeria    Africa     Northern Africa  Developing regions    80   
3    American Samoa   Oceania           Polynesia  Developing regions     0   
4           Andorra    Europe     Southern Europe   Developed regions     0   
..              ...       ...                 ...                 ...   ...   
190        Viet Nam      Asia  South-Eastern Asia  Developing regions  1191   
191  Western Sahara    Africa     Northern Africa  Developing regions     0   
192           Yemen      Asia        Western Asia  Developing regions     1   
193          Zambia    Africa      Eastern Africa  Developing regions    11   
194        Zimbabwe    Africa      Eastern Africa  Developing regions    72   

     1981  1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  \
0      39    39    47    71   340  ...  3436  3009  2652  2111  1746  1758   
1       0     0     0     0     0  ...  1223   856   702   560   716   561   
2      67    71    69    63    44  ...  3626  4807  3623  4005  5393  4752   
3       1     0     0     0     0  ...     0     1     0     0     0     0   
4       0     0     0     0     0  ...     0     1     1     0     0     0   
..    ...   ...   ...   ...   ...  ...   ...   ...   ...   ...   ...   ...   
190  1829  2162  3404  7583  5907  ...  1852  3153  2574  1784  2171  1942   
191     0     0     0     0     0  ...     0     1     0     0     0     0   
192     2     1     6     0    18  ...   161   140   122   133   128   211   
193    17    11     7    16     9  ...    91    77    71    64    60   102   
194   114   102    44    32    29  ...   615   454   663   611   508   494   

     2011  2012  2013  Total  
0    2203  2635  2004  58639  
1     539   620   603  15699  
2    4325  3774  4331  69439  
3       0     0     0      6  
4       0     1     1     15  
..    ...   ...   ...    ...  
190  1723  1731  2112  97146  
191     0     0     0      2  
192   160   174   217   2985  
193    69    46    59   1677  
194   434   437   407   8598  

[195 rows x 39 columns]

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:
    1. The full row data (all columns)
    2. For year 2013
    3. For years 1980 to 1985


In [81]:
# 1. the full row data (all columns)
df_can.loc[df_can["Country"] == "Japan"]

Country Continent        Region            DevName  1980  1981  1982  1983  \
87   Japan      Asia  Eastern Asia  Developed regions   701   756   598   309   

    1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  2012  2013  \
87   246   198  ...  1067  1212  1250  1284  1194  1168  1265  1214   982   

    Total  
87  27707  

[1 rows x 39 columns]

In [82]:
# alternate methods
df_can.iloc[87]

Country                  Japan
Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008    

In [83]:
df_can[df_can.index == 'Japan']

Empty DataFrame
Columns: [Country, Continent, Region, DevName, 1980, 1981, 1982, 1983, 1984, 1985, 1986, 1987, 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, Total]
Index: []

[0 rows x 39 columns]

In [86]:
# 2. for year 2013
df_can.loc[df_can["Country"] == "Japan", 2013]

87    982
Name: 2013, dtype: int64

In [94]:
# alternate method
# year 2013 is the last column, with a positional index of 36 (wrong it is index of 37 see next code)
df_can.iloc[87, 37]

982

In [93]:
index_2013 = df_can.columns.get_loc(2013)
print(index_2013)

37


In [96]:
# 3. for years 1980 to 1985
df_can.loc[df_can["Country"] == "Japan", [1980, 1981, 1982, 1983, 1984, 1984]]

1980  1981  1982  1983  1984  1984
87   701   756   598   309   246   246

In [97]:
# Alternative Method
df_can.iloc[87, [3, 4, 5, 6, 7, 8]]

DevName    Developed regions
1980                     701
1981                     756
1982                     598
1983                     309
1984                     246
Name: 87, dtype: object

**Exercise:** Let's view the number of immigrants from **Haiti** for the following scenarios: <br>1. The full row data (all columns) <br>2. For year 2000 <br>3. For years 1990 to 1995


In [101]:
df_can.loc[df_can["Country"] == 'Haiti']
df_can.loc[df_can["Country"] == 'Haiti', 2000]
df_can.loc[df_can["Country"] == 'Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]

1990  1991  1992  1993  1994  1995
75  2379  2829  2399  3655  2100  2014

<details><summary>Click here for a sample python solution</summary>

```python
   # 1. the full row data (all columns)
    df_can.loc['Haiti']
    #or
    df_can[df_can.index == 'Haiti']

    # 2. for year 2000
    df_can.loc['Haiti', 2000]
           
    #  3. for years 1990 to 1995
    df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]
 ```

</details>


Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index. 

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [103]:
df_can.columns = list(map(str, df_can.columns))
[print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [104]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

**Exercise:** Create a list named 'year' using map function for years ranging from 1990 to 2013. <br>Then extract the data series from the dataframe df_can for Haiti using year list. 


In [107]:
year = list(map(str, range(1990, 2014)))
df_can.loc[df_can["Country"] == "Haiti", year]

1990  1991  1992  1993  1994  1995  1996  1997  1998  1999  ...  2004  \
75  2379  2829  2399  3655  2100  2014  1955  1645  1295  1439  ...  1652   

    2005  2006  2007  2008  2009  2010  2011  2012  2013  
75  1682  1619  1598  2491  2080  4744  6503  5868  4152  

[1 rows x 24 columns]

<details><summary>Click here for a sample python solution</summary>

```python
    #The correct answer is:
    year = list(map(str, range(1990, 2014)))
    haiti = df_can.loc['Haiti', year] # passing in years 1990 - 2013
    
```
</details>


### Filtering based on a criteria <a id="4"></a>
To filter the dataframe based on a condition, we simply pass the condition as a boolean vector. 

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [108]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

0       True
1      False
2      False
3      False
4      False
       ...  
190     True
191    False
192     True
193    False
194    False
Name: Continent, Length: 195, dtype: bool


In [109]:
# 2. pass this condition into the dataFrame
df_can[condition]

Country Continent  \
0                                       Afghanistan      Asia   
8                                           Armenia      Asia   
11                                       Azerbaijan      Asia   
13                                          Bahrain      Asia   
14                                       Bangladesh      Asia   
20                                           Bhutan      Asia   
25                                Brunei Darussalam      Asia   
30                                         Cambodia      Asia   
36                                            China      Asia   
37   China, Hong Kong Special Administrative Region      Asia   
38       China, Macao Special Administrative Region      Asia   
46                                           Cyprus      Asia   
48            Democratic People's Republic of Korea      Asia   
66                                          Georgia      Asia   
79                                            India      Asia   
80                                        Indonesia      Asia   
81                       Iran (Islamic Republic of)      Asia   
82                                             Iraq      Asia   
84                                           Israel      Asia   
87                                            Japan      Asia   
88                                           Jordan      Asia   
89                                       Kazakhstan      Asia   
92                                           Kuwait      Asia   
93                                       Kyrgyzstan      Asia   
94                 Lao People's Democratic Republic      Asia   
96                                          Lebanon      Asia   
105                                        Malaysia      Asia   
106                                        Maldives      Asia   
114                                        Mongolia      Asia   
118                                         Myanmar      Asia   
121                                           Nepal      Asia   
129                                            Oman      Asia   
130                                        Pakistan      Asia   
136                                     Philippines      Asia   
139                                           Qatar      Asia   
140                               Republic of Korea      Asia   
151                                    Saudi Arabia      Asia   
156                                       Singapore      Asia   
162                                       Sri Lanka      Asia   
163                              State of Palestine      Asia   
169                            Syrian Arab Republic      Asia   
170                                      Tajikistan      Asia   
171                                        Thailand      Asia   
177                                          Turkey      Asia   
178                                    Turkmenistan      Asia   
182                            United Arab Emirates      Asia   
187                                      Uzbekistan      Asia   
190                                        Viet Nam      Asia   
192                                           Yemen      Asia   

                 Region             DevName  1980  1981  1982  1983  1984  \
0         Southern Asia  Developing regions    16    39    39    47    71   
8          Western Asia  Developing regions     0     0     0     0     0   
11         Western Asia  Developing regions     0     0     0     0     0   
13         Western Asia  Developing regions     0     2     1     1     1   
14        Southern Asia  Developing regions    83    84    86    81    98   
20        Southern Asia  Developing regions     0     0     0     0     1   
25   South-Eastern Asia  Developing regions    79     6     8     2     2   
30   South-Eastern Asia  Developing regions    12    19    26    33    10   
36         Eastern Asia  Developing regions  5123  6682  3308  1863  1527   
37         Eastern Asia

In [110]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Country Continent         Region             DevName  \
0                   Afghanistan      Asia  Southern Asia  Developing regions   
14                   Bangladesh      Asia  Southern Asia  Developing regions   
20                       Bhutan      Asia  Southern Asia  Developing regions   
79                        India      Asia  Southern Asia  Developing regions   
81   Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions   
106                    Maldives      Asia  Southern Asia  Developing regions   
121                       Nepal      Asia  Southern Asia  Developing regions   
130                    Pakistan      Asia  Southern Asia  Developing regions   
162                   Sri Lanka      Asia  Southern Asia  Developing regions   

     1980  1981  1982  1983  1984  1985  ...   2005   2006   2007   2008  \
0      16    39    39    47    71   340  ...   3436   3009   2652   2111   
14     83    84    86    81    98    92  ...   4171   4014   2897   2939   
20      0     0     0     0     1     0  ...      5     10      7     36   
79   8880  8670  8147  7338  5704  4211  ...  36210  33848  28742  28261   
81   1172  1429  1822  1592  1977  1648  ...   5837   7480   6974   6475   
106     0     0     0     1     0     0  ...      0      0      2      1   
121     1     1     6     1     2     4  ...    607    540    511    581   
130   978   972  1201   900   668   514  ...  14314  13127  10124   8994   
162   185   371   290   197  1086   845  ...   4930   4714   4123   4756   

      2009   2010   2011   2012   2013   Total  
0     1746   1758   2203   2635   2004   58639  
14    2104   4721   2694   2640   3789   65568  
20     865   1464   1879   1075    487    5876  
79   29456  34235  27509  30933  33087  691904  
81    6580   7477   7479   7534  11291  175923  
106      7      4      3      1      1      30  
121    561   1392   1129   1185   1308   10222  
130   7217   6811   7468  11227  12603  241600  
162   4547   4422   3309   3338   2394  148358  

[9 rows x 39 columns]

**Exercise:** Fetch the data where AreaName is 'Africa' and RegName is 'Southern Africa'. <br>Display the dataframe and find out how many instances are there?


In [122]:
df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')].head(5)

Country Continent           Region             DevName  1980  1981  \
160  South Africa    Africa  Southern Africa  Developing regions  1026  1118   
23       Botswana    Africa  Southern Africa  Developing regions    10     1   
119       Namibia    Africa  Southern Africa  Developing regions     0     5   
97        Lesotho    Africa  Southern Africa  Developing regions     1     1   
166     Swaziland    Africa  Southern Africa  Developing regions     4     1   

     1982  1983  1984  1985  ...  2005  2006  2007  2008  2009  2010  2011  \
160   781   379   271   310  ...   988  1111  1200  1123  1188  1238   959   
23      3     3     7     4  ...     7    11     8    28    15    42    53   
119     5     3     2     1  ...     6    19    13    26    14    16    23   
97      1     2     7     5  ...     4     0     4     1     8     7     1   
166     1     0    10     7  ...     7     7     5     6    10     3    13   

     2012  2013  Total  
160  1243  1240  40568  
23     64    76    396  
119    24    83    320  
97      0     6    107  
166    17    39    188  

[5 rows x 39 columns]

<details><summary>Click here for a sample python solution</summary>

```python
    df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]
```
</details>


### Sorting Values of a Dataframe or Series <a id="5"></a><br>
You can use the `sort_values()` function is used to sort a DataFrame or a Series based on one or more columns. <br>You to specify the column(s) by which you want to sort and the order (ascending or descending). Below is the syntax to use it:-<br><br>
```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br><br>
col_nam - the column(s) to sort by. <br>
axis - axis along which to sort. 0 for sorting by rows (default) and 1 for sorting by columns.<br>
ascending - to sort in ascending order (True, default) or descending order (False).<br>
inplace - to perform the sorting operation in-place (True) or return a sorted copy (False, default).<br>
ignore_index - to reset the index after sorting (True) or keep the original index values (False, default).<br>


Let's sort out dataframe df_can on 'Total' column, in descending order to find out the top 5 countries that contributed the most to immigration to Canada. 


In [112]:
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
top_5 = df_can.head(5)
top_5

Country Continent  \
79                                               India      Asia   
36                                               China      Asia   
183  United Kingdom of Great Britain and Northern I...    Europe   
136                                        Philippines      Asia   
130                                           Pakistan      Asia   

                 Region             DevName   1980   1981   1982   1983  \
79        Southern Asia  Developing regions   8880   8670   8147   7338   
36         Eastern Asia  Developing regions   5123   6682   3308   1863   
183     Northern Europe   Developed regions  22045  24796  20620  10015   
136  South-Eastern Asia  Developing regions   6051   5921   5249   4562   
130       Southern Asia  Developing regions    978    972   1201    900   

      1984  1985  ...   2005   2006   2007   2008   2009   2010   2011   2012  \
79    5704  4211  ...  36210  33848  28742  28261  29456  34235  27509  30933   
36    1527  1816  ...  42584  33518  27642  30037  29622  30391  28502  33024   
183  10170  9564  ...   7258   7140   8216   8979   8876   8724   6204   6195   
136   3801  3150  ...  18139  18400  19837  24887  28573  38617  36765  34315   
130    668   514  ...  14314  13127  10124   8994   7217   6811   7468  11227   

      2013   Total  
79   33087  691904  
36   34129  659962  
183   5827  551500  
136  29544  511391  
130  12603  241600  

[5 rows x 39 columns]

**Exercise:** Find out top 3 countries that contributes the most to immigration to Canda in the year 2010. <br> Display the country names with the immigrant count in this year


In [125]:
df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
top3_2010 = df_can[(df_can['Continent'] == 'Asia')][['Country', '2010']].head(3)
top3_2010

Country   2010
136  Philippines  38617
79         India  34235
36         China  30391

<details><summary>Click here for a sample python solution</summary>

```python
    df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
    top3_2010 = df_can['2010'].head(3)
    top3_2010
```
</details>


Congratulations! you have learned how to wrangle data with Pandas. You will be using alot of these commands to preprocess the data before its can be used for data visualization.


### Thank you for completing this lab!


## Author

<a href="https://www.linkedin.com/in/aklson/" target="_blank">Alex Aklson</a>


### Other Contributors
[Jay Rajasekharan](https://www.linkedin.com/in/jayrajasekharan),
[Ehsan M. Kermani](https://www.linkedin.com/in/ehsanmkermani),
[Slobodan Markovic](https://www.linkedin.com/in/slobodan-markovic),
[Weiqing Wang](https://www.linkedin.com/in/weiqing-wang-641640133/),
[Dr. Pooja](https://www.linkedin.com/in/p-b28802262/)


## Change Log


|  Date (YYYY-MM-DD) | Version | Changed By    |  Change Description                   |
|--------------------|---------|---------------|---------------------------------------|
| 2023-06-08         | 2.5     | Dr. Pooja         |  Separated from original lab        |
| 2021-05-29         | 2.4     | Weiqing Wang  |  Fixed typos and code smells.         |
| 2021-01-20         | 2.3     | Lakshmi Holla |  Changed TOC cell markdown            |
| 2020-11-20         | 2.2     | Lakshmi Holla |  Changed IBM box URL                  |
| 2020-11-03         | 2.1     | Lakshmi Holla |  Changed URL and info method          |
| 2020-08-27         | 2.0     | Lavanya       |  Moved Lab to course repo in GitLab   |




## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
